test/gateway/pollingInterval.js

```diff
@@ -0,0 +1,790 @@
+    `,
+    resolvers: resolvers,
+    federationMetadata: true,
+    subscription: true
+  })
+
+  await userService.listen(0)
+
+  const userServicePort = userService.server.address().port
+
+  gateway.register(GQL, {
+    subscription: true,
+    gateway: {
+      services: [
+        {
+          name: 'user',
+          url: `http://localhost:${userServicePort}/graphql`,
+          wsUrl: `ws://localhost:${userServicePort}/graphql`
+        }
+      ],
+      pollingInterval: 2000
+    }
+  })
+
+  await gateway.listen(0)
+
+  const ws = new WebSocket(
+    `ws://localhost:${gateway.server.address().port}/graphql`,
+    'graphql-ws'
+  )
+
+  t.equal(ws.readyState, 0)
+
+  const client = WebSocket.createWebSocketStream(ws, {
+    encoding: 'utf8',
+    objectMode: true
+  })
+
+  t.tearDown(() => {
+    client.destroy()
+  })
+
+  client.setEncoding('utf8')
+
+  client.write(
+    JSON.stringify({
+      type: 'connection_init'
+    })
+  )
+
+  client.write(
+    JSON.stringify({
+      id: 1,
+      type: 'start',
+      payload: {
+        query: `
+          subscription {
+            updatedUser {
+              id
+              name
+            }
+          }
+        `
+      }
+    })
+  )
+
+  client.on('data', (chunk) => {
+    const data = JSON.parse(chunk)
+
+    if (data.type === 'connection_ack') {
+      gateway.inject({
+        method: 'POST',
+        url: '/graphql',
+        body: {
+          query: `
+          mutation {
+            triggerUser
+          }
+        `
+        }
+      })
+    } else if (data.type === 'data' && data.id === 1) {
+      const { payload: { data: { updatedUser = {} } = {} } = {} } = data
+
+      t.deepEqual(updatedUser, {
+        id: 'u1',
+        name: 'John'
+      })
+    }
+  })
+
+  await once(client, 'data')
+  await once(client, 'data')
+
+  userService.graphql.replaceSchema(
+    buildFederationSchema(`
+      extend type Query {
+        me: User
+      }
+
+      extend type Subscription {
+        updatedUser: User
+      }
+
+      extend type Mutation {
+        triggerUser: Boolean
+      }
+
+      type User @key(fields: "id") {
+        id: ID!
+        name: String!
+        lastName: String
+      }
+    `)
+  )
+
+  userService.graphql.defineResolvers(resolvers)
+
+  await clock.tickAsync(2000)
+  await clock.tickAsync()
+
+  t.deepEqual(Object.keys(gateway.graphql.schema.getType('User').getFields()), [
+    'id',
+    'name',
+    'lastName'
+  ])
+
+  client.on('data', (chunk) => {
+    const data = JSON.parse(chunk)
+
+    if (data.id === 1) {
+      const { payload: { data: { updatedUser = {} } = {} } = {} } = data
+
+      t.deepEqual(updatedUser, {
+        id: 'u1',
+        name: 'John'
+      })
+    }
+  })
+
+  await gateway.inject({
+    method: 'POST',
+    url: '/graphql',
+    body: {
+      query: `
+          mutation {
+            triggerUser
+          }
+        `
+    }
+  })
+
+  await once(client, 'data')
+
+  t.equal(ws.readyState, 1)
```

Comment:
I think the first client should be closed now.

Reply:
I created a new client for getting a new subscription.
I can manually close ws.client inside tests, but we have it inside `t.tearDown()`.

If we compare it with an end-user perspective.

For example,
You're on the page with notifications. You're using an application and getting new notifications in real-time. Some developer updated GraphQL Type for notifications (he added one field).

We can close a WebSocket connection. In this case, end-user has to refresh the page after these changes.
If we don't close WebSocket connection, end-user can use an application as before and in the future, he will refresh the page and get new data.

Do you agree with me?

Reply:
If you change the schema, there is a very good chance that all the connections to that peer will crash/return bad data :/. I think it's better if the client is closed.

Reply:
@mcollina 
Could you tell me, how to close a current WebSocket connection?
I didn't find an interface for it :cry:

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/close or `stream.destroy()`.

Reply:
@mcollina 

I tried to close websocket connection and it works but I can't test it.
Could you take a look at the current implementation of it? Is it correct?